BattleSimulator - Calculate and show standard deviation

emmanuel-perlow opened the pull request:
Calculate and show standard deviation in battle simulator.

I used String.format() which is a printf like function instead of the DecmialFormat class since this makes it easier to see the whole string "a ± b / c" in the code. (According to the docs String.format() is also local sensitive.)

## Testing

No tests where written as I have no clue how to write those. (Might also not be necessary.)

## Additional Notes to Reviewer

I named the functions calculating the standard deviation "getXXXStandardDeviation()" not very happy with that name but fits. Feel free to change that as you like.

## Release Note

<!--RELEASE_NOTE-->NEW|Battle Calculator now shows the standard deviation along with the expected value.<!--END_RELEASE_NOTE-->

## Fixes

Fixes #9489

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/triplea-game/triplea?pullRequest=9490) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/triplea-game/triplea?pullRequest=9490) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/triplea-game/triplea?pullRequest=9490) it.</sub>

DanVanAtta commented on game-app/game-core/src/main/java/games/strategy/triplea/odds/calculator/AggregateResults.java:
```diff
@@ -141,6 +220,26 @@ public double getAverageTuvSwing(
+    if (results.isEmpty()) {
+      return 0.0;
+    }
+    final double mu = getAverageAttackingUnitsLeftWhenAttackerWon();
```
`mu` might be hard to follow as a variable name for the casual reader.  `AverageUnits` might be more clear.
Better yet, I think the standard deviation computation can be skipped altogether, this lib looks to have some promising APIs: https://commons.apache.org/proper/commons-math/javadocs/api-2.2/org/apache/commons/math/stat/descriptive/moment/StandardDeviation.html#evaluate(double[],%20double)

emmanuel-perlow replied:
Sure we can use the lib.  The immediate question is: Should we also replace the average calculation with that lib?
(remark: in our case "average", "expected value", "first momentum" and "mean" all mean the same [pun intended ;-) ]).

Looking at the lib, there is an overhead (object creation, function calls) involved when using it to compute the mean in contrast to calculating it directly.  I assume the AI calls (and thus calculates) the expected value a lot, so this might become a performance regression (though I have not tested it).

DanVanAtta replied:
As of Java 8, invoking functions and creating objects are exceedingly cheap. It can be faster to create an object than it is to invoke a no-op function. Furthermore, those are NS scale operations, there are far greater overheads.

Otherwise, yeah, it's a great idea to also replace the computation for 'average' as well. 'Expected value' is not necessarily average, is it? (We are at risk of starting a mean set of puns)

DanVanAtta commented on game-app/game-core/src/main/java/games/strategy/triplea/odds/calculator/AggregateResults.java:
```diff
@@ -170,20 +284,66 @@ public double getAverageTuvSwing(
+    // Winning is a Bernoulli Experiment with propability p = getAttackerWinPercent(),
+    // thus the variance is p(1-p).
+    // [Or do the math yourself: w = number of wins, n = num of battles, p = w/n
+    // Var = (w*(1-p)^2 + (n-w)*(0-p)^2)/n = ... = p(1-p)]
```
Do you happen to know this formula by heart? A link for more information would seemingly be helpful.

emmanuel-perlow replied:
Ohm ... yeah ... but I agree a reference is nice.

DanVanAtta commented on game-app/game-core/src/main/java/games/strategy/triplea/odds/calculator/AggregateResults.java:
```diff
@@ -170,20 +284,66 @@ public double getAverageTuvSwing(
     return results.stream().filter(BattleResults::attackerWon).count() / (double) results.size();
   }
 
+  /** Returns the standard deviation of the attacker wins percentage. */
```
The std deviation for number of times the battle is won is seemingly very helpful. This is a good addition :+1:

emmanuel-perlow replied:
Funny, I think that this std dev is the least useful of all the added ones (I even considered not adding it) because it only depends on the average (=win percentage), thus does not add any information.  For TUV swing etc the std dev really adds something not already covered by the expected value/average.

DanVanAtta commented on game-app/game-core/src/main/java/games/strategy/triplea/odds/calculator/AggregateResults.java:
```diff
@@ -197,13 +357,44 @@ public double getAverageBattleRoundsFought() {
+    if (results.isEmpty()) {
+      return 0.0;
+    }
+    // Drawing is a Bernoulli Experiment with propability p = getDrawPercent(),
```
This seems repetitious, a good candidate to extra to a lib.

emmanuel-perlow replied:
Do you mean outsourcing the line `return p*(p-1)`? The important (and hard) part here is to realise that (in this case) drawing is a Bernoulli Experiment. This observation has to be redone in all three cases (win, lose, draw) and IMHO cannot be outsourced into a lib.  The remaining line `p*(p-1)` is so small that it does not seem worth it having a function
`getExpectedValueForBernoulliExperiment(double p)` (or `new BinomialDistribution(p).getExpectedValue()`).

DanVanAtta replied:
Overall it should be simpler extracted to a method. The method name can state "computeBernoulliExperimentExpectedValue(..)", which then removes the need for the comment at the site of usage. Secondarily the logic and documentation becomes consolidated. Last, it allows for you to test that component in isolation if you were to choose to do so (and arguably should).

DanVanAtta reviewed (commented):
Some initial comments. So far:
- std deviation can be done through a lib, no need to compute it ourselves
- if we just invoke new library code to get some more numbers, there is nothing to test. The existing code and any new computations though would OTOH be readily testable. We can discuss this further
- the UX & UI updates still need to be reviewed and potentially discussed

DanVanAtta commented:
Thank you for taking a crack at updating the source. Let me know which questions you may have. There are some more things to do here:
- fix automated checks / the build
- update PR appropriately to respond to the initial review
- review further and finish code reviewing (for me to do)

Cernelius commented:
While at it, can you please have the standard deviation and all the rest (especially the win probability) showing rounded to 1 or 2 decimals instead of rounded to natural?

Reference:
https://forums.triplea-game.org/topic/2604/less-approximative-battle-calculator

emmanuel-perlow commented:
Hi @Cornelius

> While at it, can you please have the standard deviation and all the rest (especially the win probability) showing rounded to 1 or 2 decimals instead of rounded to natural?

I have no objection against this but I will not implement this in this pull request.  Once it's finished I'll make a new one with your request.
Be aware, however, that for the additional digits to be meaning full, you must do more than 10,000 rounds of simulation!

Cernelius commented:
> Hi @Cernelius
> 
> > While at it, can you please have the standard deviation and all the rest (especially the win probability) showing rounded to 1 or 2 decimals instead of rounded to natural?
> 
> I have no objection against this but I will not implement this in this pull request. Once it's finished I'll make a new one with your request.
> Be aware, however, that for the additional digits to be meaning full, you must do more than 10,000 rounds of simulation!

I've opened a feature request in the proper place, namely:
https://forums.triplea-game.org/topic/2604/less-approximative-battle-calculator

I'll comment there.

emmanuel-perlow commented:
Updated the code to use the commons-math library by Apache as suggested by @DanVanAtta.

Other stuff:
* Added Javadoc to (almost) all methods
* Converted average function to also use commons-math
* Moved `getDrawPercantage()` to `getXXXWinPercentage()`

DanVanAtta commented on game-app/game-core/src/main/java/games/strategy/triplea/odds/calculator/AggregateResults.java:
```diff
@@ -13,30 +13,79 @@
+ * <p>This class does not restrict the battle result to come from the same battle setup. If this is
+ * desired, the user must ensure that this property hold.
+ *
+ * <p>Note on bias correction: Given a set of data points there are different variances (or standard
```
Nit: Please check if you can shorten up this javadoc with a second edit.
EG: "For more information see any book about statistics or, can just be shorted to", "see: <url>"

Another point of view, see how many words you can take out while still retaining the original meaning. Ditto for any commentary and blocks of text added.

Site-note: great to see the documentation added :+1:

DanVanAtta commented on game-app/game-core/src/main/java/games/strategy/triplea/odds/calculator/AggregateResults.java:
```diff
@@ -13,30 +13,79 @@
     this.results.addAll(results);
   }
 
+  /**
```
nit, this block can be replaced with a lombok `@Getter` on the variable, then we don't need to worry about a javadoc or writing the 'get' method by hand either.

emmanuel-perlow replied:
Will do.  However, I noticed you squash all commits into one, so I'll postpone this into another pull request – probably together with the javadoc update – as they are not related to implementing the standard deviation.

DanVanAtta commented on game-app/game-core/src/main/java/games/strategy/triplea/odds/calculator/AggregateResults.java:
```diff
@@ -107,103 +195,362 @@ public double getAverageTuvSwing(
+    final IntegerMap<UnitType> defenderCostsForTuv = TuvUtils.getCostsForTuv(defender, data);
+    // The 'true' argument means calculate the variance using bias correction.
+    final StandardDeviation stdDev = new StandardDeviation(true);
+    final double s =
+        stdDev.evaluate(
+            results.stream()
+                .mapToDouble(
+                    result ->
+                        TuvUtils.getTuv(result.getRemainingAttackingUnits(), attackerCostsForTuv)
+                            - TuvUtils.getTuv(
+                                result.getRemainingDefendingUnits(), defenderCostsForTuv))
+                .toArray());
+    return Double.isNaN(s) ? 0 : s;
```
This block is slow to grok. Would you mind looking to see if it can be made 'lighter' to understand?
Suggestions:
- spell out the variable 's'
- put a comment above the stream operation to give a short summary of what we're doing
- consider using intermediate variables so there is less computation per line of code

emmanuel-perlow replied:
Let's see what we can do.  The sad part is, that the auto formatting breaks the symmetry between attacker TUV and defender TUV, otherwise you would instantly see that the middle lines are the same but with attacker changed into defender.

DanVanAtta commented on game-app/game-core/src/main/java/games/strategy/triplea/odds/calculator/AggregateResults.java:
```diff
@@ -107,103 +195,362 @@ public double getAverageTuvSwing(
-    return count / total;
+  /**
+   * Returns the average number of attacking units surviving the battles.
+   *
+   * @return the average of surviving attacking units
```
nit, if the main paragraph of a javadoc repeats the `@return` part, typically I'll leave the `@return` off. A tell-tale sign is when the comment begins with the word 'returns'. We do not generate javadocs in this project and it's super rare for anyone to view the rendered javadocs in any context, hence the target audience is really another Dev, in an IDE reading this with the code right below it.

DanVanAtta commented on game-app/game-core/src/main/java/games/strategy/triplea/odds/calculator/AggregateResults.java:
```diff
@@ -107,103 +195,362 @@ public double getAverageTuvSwing(
+   * @return the average of surviving attacking units
+   */
+  public double getAverageAttackingUnitsLeft() {
+    final Mean mean = new Mean();
+    final double m =
+        mean.evaluate(
+            results.stream()
+                .map(BattleResults::getRemainingAttackingUnits)
+                .mapToDouble(Collection::size)
+                .toArray());
+    // If there are no battles, mean.evaluate() returns Double.NaN.
+    return Double.isNaN(m) ? 0 : m;
```
I wonder if this can be made a bit easier to understand. As a suggestion, do you think something like this would be easier to follow (or just all in the eye of the beholder?)?
```
Double[] attackingUnitsLeft = results.stream()
                .map(BattleResults::getRemainingAttackingUnits)
                .mapToDouble(Collection::size)
                .toArray();
return attackingUnitsLeft.length == 0 ? 0 : new Mean(attackingUnitsLeft);
```
This avoids needing to use the `isNan` or comment that `mean` returns that value for empty arrays, it's more explicit in the above. We also reduce the number of intermediate variables as well. I suppose a similar pattern could apply to the other methods as well. WDYT?

emmanuel-perlow replied:
What I don't like at the suggested code is that we use the "check first" pattern, a pattern which is often discourage (in particular when doing I/O) better put the check afterwards which is summed up in the mantra "Better ask for forgiveness than permission". So, one should aim for the "check afterwards" pattern (if the library API allows for this), where the return value of the actual function is checked.  Then and only then, we can be sure that a function call worked/didn't work.

Applied to the above code, this means that Mean.evaluate() could still return NaN.  (But looking at some version of code of the library, I do not see any case besides no input data, where this could happen).
If we use the NaN check afterwards we are assured to have caught every possible outcome and never return NaN in our function.  However, the above is really nitpicking for such a simple API usage.

Personally, I like my version more (not a surprise) but I also have a reason for it:
In that version the (core of the) algorithm is complete in the line `double s = mean.evalute(...)`.  Then afterwards it follows the check for rare corner cases. In the proposed version these two things are mixed, first we do stuff to generate an array, then we check the corner cases, then we finish the calculation of the mean. If I'm not familiar with a code base, I'd like first to know what do we do there and then afterwards understand (or see) what other cases have additionally be taken care of (e.g. error handling, corner cases, ...).

I believe in the end it doesn't really matter both are reasonably fine.


After contemplating a little more.  I think we should even go a similar way as the library and return NaN in case we cannot calculate the requested quantity.  Do we really gain something by returning 0 if we cannot calculate the mean? Or would it better to return NaN to tell the caller "sorry, not valid what you trying to do".

I even wonder whether we should do the NaN check at all.  Do we really gain something by return 0 when the reality is "sorry we cannot calculate the mean/stddev".  In that case, NaN is a valid return value indicating "something's wrong with your input data". (Usually you would throw an exception, however, for the `getAverageAttackingUnitsLeftWhenAttackerWon()` the caller cannot reasonable ensure that there is a won battle.)

Deciding so would make the NaN checks redundant and the only change would be that the user gets a `NaN ± NaN / 8` for such cases, that however we can circumvent with a check when formatting that number and print `N/A` instead. (There is no other call to the `getAverageAttackingUnitsLeftWhenAttackerWon()` function.)
What do you think about this proposal?

DanVanAtta replied:
Good analysis. Returning and displaying `N/A` seems reasonable.

DanVanAtta reviewed (commented):
Great to see the lib put in place. I got about 70% of the way through reviewing, left some feedback for you to consider.

DanVanAtta commented:
![Screenshot from 2021-08-19 00-35-06](https://user-images.githubusercontent.com/12397753/130027490-fe1a0fde-2072-484a-82cc-fd2c23b73d39.png)

@emmanuel-perlow I've checked this some more and took the above screenshot. I think this is headed in a good direction.

One observation, a standard deviation that gives a win rate above 100% is odd. Perhaps we should not give a standard deviation for win/loss and instead give a raw number and the total. EG:

A)
```
Attacker wins: 23 (23%)
Draw:  2  (2%)
Defender wins: 75 (75%)
```

B)
```
Attacker wins: 23% (23)
Draw:  2%  (2)
Defender wins: 75% (75)
```

Second, the `/8` in the 'remaining' statistics struck me as odd at first glance. I think it is because it reads like: `a ± (b / c)`
Do you have any thoughts on how to fix this? I think writing it as `a/c ± b` could help and reducing the decimals on `b` (or perhaps overall)

emmanuel-perlow commented:
> One observation, a standard deviation that gives a win rate above 100% is odd. Perhaps we should not give a standard deviation for win/loss and instead give a raw number and the total. [...]

Giving the number of wins and the same number in percent, is not very useful IMHO as they are (up to a scaling factor) the same. I'm also almost certain that most people appreciate percentage more than have to guess how high (or low) a 1758 out of 2138 is.

For the original concern: You hit a common misconception.  The standard deviation in `a ± b` does **not** say: "The result lies within `a-b` and `a+b`" (although the notation [which is very common] highly suggests that).  It says more something like:  "there is an error and that error is on average about `b`".  We cannot say anything about the error. The actual error might be larger or smaller than `b` this is not ruled out nor is anything said about how likely a particular value (or range of values) for the error is.  In particular in your screenshot it might be that the real probability for winning (i.e. the probability if you calculate the battle and not simulate it) is below 84%-36% = 48% though you had to be pretty pretty unlucky with your dice rolls.  If you really want to spice things up, you can go and calculate some confidence intervals to some confidence level but that will almost nobody understand and also cannot be written so concise as `a ± b` .

> Second, the /8 in the 'remaining' statistics struck me as odd at first glance. I think it is because it reads like: a ± (b / c)
Do you have any thoughts on how to fix this? I think writing it as a/c ± b could help and reducing the decimals on b (or perhaps overall)

I see what you mean and I'm also unhappy with the state of affairs.  During development I also tested your suggested `a/c ± b` but found that even more awkward because if splitting at `/` it adds the uncertainty to the max number `c ± b` and if splitting at `±` it's wrong as it is not the uncertainty for `a/c`.  I also tried `a/c ± b/c` which would be correct but that was to redundant.  I also thought about `(a ± b) / c` but I wasn't sure about starting the answer with a parenthesis (though that would have been my second choice).  I guess preferred would be to visually separate the `c` from the rest similar to a table (think Excel, or Calc), but that would require a lot of new code.  Maybe replacing `/` with `of` would help  (`0.40 ± 1.11 of 8`)?

tvleavitt commented:
Emmanuel,

If we implement "(0.40 ± 1.11 of 8)", can we get a clear indication of the
"order of operations" here? Such as "0.40 +/- 1.11, of 8"? I'm honestly not
even sure what that means (pardon my ignorance)... is it a range of 0.40 -
1.11 to 0.40 + 1.11, i.e. anywhere from -0.71 to 1.51 of 8?

Thomas

On Thu, Aug 19, 2021 at 10:35 AM emmanuel-perlow ***@***.***>
wrote:

> One observation, a standard deviation that gives a win rate above 100% is
> odd. Perhaps we should not give a standard deviation for win/loss and
> instead give a raw number and the total. [...]
>
> Giving the number of wins and the same number in percent, is not very
> useful IMHO as they are (up to a scaling factor) the same. I'm also almost
> certain that most people appreciate percentage more than have to guess how
> high (or low) a 1758 out of 2138 is.
>
> For the original concern: You hit a common misconception. The standard
> deviation in a ± b does *not* say: "The result lies within a-b and a+b"
> (although the notation [which is very common] highly suggests that). It
> says more something like: "there is an error and that error is on average
> about b". We cannot say anything about the error. The actual error might
> be larger or smaller than b this is not ruled out nor is anything said
> about how likely a particular value (or range of values) for the error is.
> In particular in your screenshot it might be that the real probability for
> winning (i.e. the probability if you calculate the battle and not simulate
> it) is below 84%-36% = 48% though you had to be pretty pretty unlucky with
> your dice rolls. If you really want to spice things up, you can go and
> calculate some confidence intervals to some confidence level but that will
> almost nobody understand and also cannot be written so concise as a ± b .
>
> Second, the /8 in the 'remaining' statistics struck me as odd at first
> glance. I think it is because it reads like: a ± (b / c)
> Do you have any thoughts on how to fix this? I think writing it as a/c ± b
> could help and reducing the decimals on b (or perhaps overall)
>
> I see what you mean and I'm also unhappy with the state of affairs. During
> development I also tested your suggested a/c ± b but found that even more
> awkward because if splitting at / it adds the uncertainty to the max
> number c ± b and if splitting at ± it's wrong as it is not the
> uncertainty for a/c. I also tried a/c ± b/c which would be correct but
> that was to redundant. I also thought about (a ± b) / c but I wasn't sure
> about starting the answer with a parenthesis (though that would have been
> my second choice). I guess preferred would be to visually separate the c
> from the rest similar to a table (think Excel, or Calc), but that would
> require a lot of new code. Maybe replacing / with of would help (0.40 ±
> 1.11 of 8)?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/triplea-game/triplea/pull/9490#issuecomment-902109537>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABB5HALFUCV2WZKL37ROOIDT5U6FXANCNFSM5B4RTWXA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>


-- 
Thomas Leavitt
Internet enabled since 1990

emmanuel-perlow commented:
Hi Thomas

Thanks for dropping by and giving your feedback.

> If we implement "(0.40 ± 1.11 of 8)", can we get a clear indication of the "order of operations" here? Such as "0.40 +/- 1.11, of 8"?

We would implement that without the parenthesis, i.e. `0.40 ± 1.11 of 8`. It's hard to spot but is your suggestion to add a comma before "of", i.e. `0.40 ± 1.11, of 8`?

> I'm honestly not even sure what that means (pardon my ignorance)... is it a range of 0.40 - 1.11 to 0.40 + 1.11, i.e. anywhere from -0.71 to 1.51 of 8?


In this context (that is statistics) `x ± s` it is *not* a range, nor is it two values (like in the quadratic formula).
It describes two (more or less) independent numbers:  the value `x` and the *uncertainty* `s` for it.  If all our data points (the results of the simulated battles) are very close to `x`, then `s` is small, if the results vary a lot, then `s` is larger. In science such statistics have to be done very often and a common and well established estimate for the uncertainty is the standard deviation, often denoted `σ` (sigma; the Greek version of the letter `s`) and is usually written as `x ± s`.
As a rule of thumb (doing some approximations, simplifications, and assumptions) we can say that about 68% of the data points (i.e. the battle results) lie within the interval `[x-s,x+s]` (i.e. between `x-s` and `x+s`).  This is also true for any future simulations we do:  If we do another 100 battle simulations, then about 68% of the results will be in that interval – be aware though this is again random so it might be more or might be less.

What we **cannot** conclude is

* that the true value of the property `x` lies within the interval `[x-s,x+s]`
* that the true value of the property `x` lies with 68% certainty in `[x-s,x+s]`
* when we rerun the experiment (i.e. fight the battle) that the result lies in `[x-s,x+s]`

**What does that mean for your every day live of conquering the world?** It means there is a very real chance that your TUV swing/attackers left/... is not "the rounded version of `x`" but will be something between `x-s` and `x+s`.
Moreover, the chance to have a result outside of that interval is small and can be ignored.

You may want to read the introduction examples[1] and the beginning of the section about interpretation and application[2] of the Wikipedia article about the standard deviation (maybe in your first language as the topic becomes fast very subtle) .

In any case the bottom line (and a lesson for life) is this: `s` is an uncertainty and is not directly added/subtracted to/from `x`.

Hope this makes it clearer.

[1] https://en.wikipedia.org/wiki/Standard_deviation#Basic_examples
[2] https://en.wikipedia.org/wiki/Standard_deviation#Interpretation_and_application

Cernelius commented:
Especially when playing dice, a standard deviation value is scarcely meaningful when the average difference from the average for the results which are greater than the average is different from the same for those which are lesser than the average (meaning that you have an average deviation upwards which is different from the average deviation downwards). Practically, knowing the standard deviation is insightful only as long as you are looking at a Gaussian distribution (which is usually not the case in the matter at hand). For example, in the image posted by @DanVanAtta, the "Ave. Defender Units Left", given as 0.40 with standard deviation 1.11, clearly implies we are looking at a distribution that is far from being Gaussian, in which the average deviation for upwards results is much greater than 1.11 (and, of course, all downwards results deviate exactly by 0.40 instead of 1.11).

As much as I commend the intent behind it, I'm not sure this new feature generally adds enough value to the average player to justify its existence as I see it being fairly worthwhile as an information only for the average TUV swing. It might do so if also a median value is added but only for persons which are skilled in reading statistics (so probably only a small minority of players).

Besides, I actually see no value in having a standard deviation for "Attacker Wins", "Draw" and "Defender Wins". I'm not even sure what the standard deviation even means there.

emmanuel-perlow commented:
> Especially when playing dice, a standard deviation value is scarcely meaningful when the average difference from the average for the results which are greater than the average is different from the same for those which are lesser than the average (meaning that you have an average deviation upwards which is different from the average deviation downwards). Practically, knowing the standard deviation is insightful only as long as you are looking at a Gaussian distribution (which is usually not the case in the matter at hand).

I think that's missing the point.  The standard deviation should not determine the distribution of the simulated results or describe it completely (like it would for a normal distribution) but instead should give some additional insight on how the results are distributed, e.g. if the simulation results vary only a little bit or stark (translated into the game this would mean a more safe strat vs a more risky strat).  It also should indicate that there is variance (both the variance by the random experiment and the variance due to the simulation) and that the average is not what you often will observe.

There are many many ways of estimating and describing this, some are more accurate than others but usually more complex, others apply only to special circumstance, etc.  And sure you can be more precise than given the standard deviation, e.g. by given a (non-symmetric) confidence interval.  However, the simplicity of using the standard deviation is that it is
a) a simple number,
b) well established,
c) well understood, and
d) easy to calculate.

The intention of this number is to provide more like a feeling to the player about the variance of the results, i.e. describing it more qualitatively than quantitatively.  It's not about providing a thorough statistical analysis of how likely some event under some restriction is,  it's more about: ok this battle swings around plus/minus two additional units lost, can I afford that.  And the magic is that the standard deviation gives a number which roughly describes that.

> For example, in the image posted by @DanVanAtta, the "Ave. Defender Units Left", given as 0.40 with standard deviation 1.11, clearly implies we are looking at a distribution that is far from being Gaussian, in which the average deviation for upwards results is much greater than 1.11 (and, of course, all downwards results deviate exactly by 0.40 instead of 1.11).

Not every number we calculate adds equal useful information, some give more some give less depending on the situation.  This is already true for the provided averages, e.g. in the screenshot the win rate for the attackers is 84% and the *avg defender units left* is 0.40 but that adds nothing of value since we already know we lose a lot, i.e. 0 units are left.  The value to look for in this case would be the *avg defender unts left if lost* and there we have 2.78 ± 1.41 a value which adds some new information (both the average and the standard deviation).  Another example for the standard deviation is the *attacker units left if won* with 4.89 ± 2.03, here variance indicates that it is to be expected to lose the (more expensive) swordman and an axeman and that in contrast it is quite unlikely to lose a swordman and two axeman.  Again, this should not be taken as exact probabilities but as a feeling (here about plus minus two units).

Lastly, although most (maybe even all) players will not interpret the numbers statically correct (even scientists doing this in their everyday work don't), they do interpret them intuitively in the right way.  More importantly, they do not misinterpret them in any harmful way an example can be seen three posts above.

Puhh, that was more than I wanted to write.

TL;DR: its more about giving a feeling for the variance of the data than a precise description of the data

P.S. I'm currently even thinking of providing somehow the histogram of the results, so the player can really see the distribution, e.g. in a tool tip or separate dialog.

DanVanAtta commented:
Leaving a simple percentage for win vs loss is okay with me. The std
deviation there did not make a lot of sense. I think it might be because
the std deviation is over only two values, 0 and 1, but makes a lot more
sense when there is a greater range (eg, number units remaining)

For the "/c", we should consider removing it, the total unit count is
already displayed.

For the std deviation on units left, why only one std deviation and say not
two?

On Fri, Aug 20, 2021, 11:56 AM emmanuel-perlow ***@***.***>
wrote:

> Especially when playing dice, a standard deviation value is scarcely
> meaningful when the average difference from the average for the results
> which are greater than the average is different from the same for those
> which are lesser than the average (meaning that you have an average
> deviation upwards which is different from the average deviation downwards).
> Practically, knowing the standard deviation is insightful only as long as
> you are looking at a Gaussian distribution (which is usually not the case
> in the matter at hand).
>
> I think that's missing the point. The standard deviation should not
> determine the distribution of the simulated results or describe it
> completely (like it would for a normal distribution) but instead should
> give some additional insight on how the results are distributed, e.g. if
> the simulation results vary only a little bit or stark (translated into the
> game this would mean a more safe strat vs a more risky strat). It also
> should indicate that there is variance (both the variance by the random
> experiment and the variance due to the simulation) and that the average is
> not what you often will observe.
>
> There are many many ways of estimating and describing this, some are more
> accurate than others but usually more complex, others apply only to special
> circumstance, etc. And sure you can be more precise than given the standard
> deviation, e.g. by given a (non-symmetric) confidence interval. However,
> the simplicity of using the standard deviation is that it is
> a) a simple number,
> b) well established,
> c) well understood, and
> d) easy to calculate.
>
> The intention of this number is to provide more like a feeling to the
> player about the variance of the results, i.e. describing it more
> qualitatively than quantitatively. It's not about providing a thorough
> statistical analysis of how likely some event under some restriction is,
> it's more about: ok this battle swings around plus/minus two additional
> units lost, can I afford that. And the magic is that the standard deviation
> gives a number which roughly describes that.
>
> For example, in the image posted by @DanVanAtta
> <https://github.com/DanVanAtta>, the "Ave. Defender Units Left", given as
> 0.40 with standard deviation 1.11, clearly implies we are looking at a
> distribution that is far from being Gaussian, in which the average
> deviation for upwards results is much greater than 1.11 (and, of course,
> all downwards results deviate exactly by 0.40 instead of 1.11).
>
> Not every number we calculate adds equal useful information, some give
> more some give less depending on the situation. This is already true for
> the provided averages, e.g. in the screenshot the win rate for the
> attackers is 84% and the *avg defender units left* is 0.40 but that adds
> nothing of value since we already know we lose a lot, i.e. 0 units are
> left. The value to look for in this case would be the *avg defender unts
> left if lost* and there we have 2.78 ± 1.41 a value which adds some new
> information (both the average and the standard deviation). Another example
> for the standard deviation is the *attacker units left if won* with 4.89
> ± 2.03, here variance indicates that it is to be expected to lose the (more
> expensive) swordman and an axeman and that in contrast it is quite unlikely
> to lose a swordman and two axeman. Again, this should not be taken as exact
> probabilities but as a feeling (here about plus minus two units).
>
> Lastly, although most (maybe even all) players will not interpret the
> numbers statically correct (even scientists doing this in their everyday
> work don't), they do interpret them intuitively in the right way. More
> importantly, they do not misinterpret them in any harmful way an example
> can be seen three posts above.
>
> Puhh, that was more than I wanted to write.
>
> TL;DR: its more about giving a feeling for the variance of the data than a
> precise description of the data
>
> P.S. I'm currently even thinking of providing somehow the histogram of the
> results, so the player can really see the distribution, e.g. in a tool tip
> or separate dialog.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/triplea-game/triplea/pull/9490#issuecomment-902892405>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC6SZOMQQDZCFJCL3HTMTVTT52QOLANCNFSM5B4RTWXA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>

DanVanAtta commented:
For normal distributions, the average plus or minus two times two standard deviations will contain 95% of all data points. Hence, if the distribution is normal (law of large numbers tends to make that true), then reporting a range of two std deviations will capture most data points.

Putting that to the side, we do need to think how we can move this forward. The problems to solve:
- `/c` looks off
- The plus/minus number might not be immediately clear and thus not helpful for what exactly it even is

You mentioned about displaying a distribution, and I agree that would be the most beneficial. We could maybe do something like have a 'details' button that opens a new window that could show a histogram. I do wonder if that is actually overkill and would be more time consuming than it is worth it. Though, given a distribution is arguably the most useful, a simple percentage for win/loss is exactly that, so we do not need a std deviation on win/loss/draw counts.

I think simply removing the `/c` is a fine option, the UI could use some re-organization and improvements overall, better to reduce redundant data if it'll make the display more clear. That though leaves the problem of clearly indicating to the user they are looking at a standard deviation. Maybe displaying a 'min/max' range might be more clear and be just as useful? I think we need to consider more ideas how to more clearly convey what the numbers are. A 'help' button could be placed somewhere in the battle calculator to help give more detailed meaning to the numbers, though the display should still keep it quite obvious what we are showing.

Maybe it might make sense to remove all of the new standard deviation calculations from this PR so we can then merge it to gain benefit from the new javadocs and the consolidated calculations. Then follow-up with a second PR that adds std-deviations (or some other treatment) in which we could take our time (having large code diffs hanging out always makes me nervous). Breaking this is up is not mandatory, but if you're inclined to do so it could be helpful to let some code land sooner while we work on the UX problems together.

emmanuel-perlow commented:
> You mentioned about displaying a distribution, and I agree that would be the most beneficial. We could maybe do something like have a 'details' button that opens a new window that could show a histogram. I do wonder if that is actually overkill and would be more time consuming than it is worth it. Though, given a distribution is arguably the most useful, a simple percentage for win/loss is exactly that, so we do not need a std deviation on win/loss/draw counts.
>
> [...] the UI could use some re-organization and improvements overall, better to reduce redundant data if it'll make the display more clear.

If you give me some leeway, I'll do a complete redesign of the battle calculator.  I also have some other ideas which I wanted to implement after this PR (among others: histogram, allowing to add multiple defenders). Starting with a blank sheet would allow more freedom as those don't have to fit within the current GUI design.  I'll ask because redesigns usually experience resistance of "but its not like I'm used to", if this wouldn't have a fair chance of being considered/merged, I wouldn't invest my time (sorry for being so blunt).

> Maybe it might make sense to remove all of the new standard deviation calculations from this PR so we can then merge it to gain benefit from the new javadocs and the consolidated calculations.

Sure, I can break this PR up.  However, there is still the outstanding (mini)issue with the NaN value (even for the mean), see my comment and proposal: https://github.com/triplea-game/triplea/pull/9490#discussion_r692346700.

DanVanAtta commented:
Sorry to not have gotten back sooner. @emmanuel-perlow , do feel free to '@' mention me as a reminder if something is really needing feedback.

A bigger redesign of the battle calculator would be well called for. After this next release, we are contemplating major data model changes and so we would want to coordinate. 

I think the outstanding questions are:
- how to deal with N/A (I'd agree to just display it to user)
- how to deal with the `/c` (IMO just remove `/c` for now)
- showing variance, do we instead show min-max? Would that be more helpful than variance.

That last question I really am not sure, it could be very hard for the naive user to know exactly what the plus/minus would mean. This would be a good question to bring to the user-forums.

emmanuel-perlow commented:
> A bigger redesign of the battle calculator would be well called for. After this next release, we are contemplating major data model changes and so we would want to coordinate.

I was more thinking about an UI overhaul but it's good to know that the ground might change under the feet.

As a side question: When you talk about "this next release" (I guess it's 2.6), are we talking weeks, months, or years?

> *   showing variance, do we instead show min-max? Would that be more helpful than variance.
>
> That last question I really am not sure, it could be very hard for the naive user to know exactly what the plus/minus would mean. This would be a good question to bring to the user-forums.

Would you be so kind and point me to the correct subforum?

beelee1 commented:
@emmanuel-perlow  Dan may have a better answer,  but this is probably the subforum you want https://forums.triplea-game.org/category/42/feature-requests-ideas
here's an old topic 
https://forums.triplea-game.org/topic/1944/development-discussion-speeding-up-battle-calculator-and-thus-hard-ai?_=1631217166980
but you might want to just start a new one. Idk : )

DanVanAtta commented:
I'm going to close this PR for now.

I like where this update is going though it is really difficult to determine a very good UX. Let's re-open this if we can come up with the right UX.
